README.md
# tkit-mp-openapi-plugin

tkit microprofile openapi plugin 
[![License](https://img.shields.io/badge/license-Apache--2.0-green?style=for-the-badge&logo=apache)](https://www.apache.org/licenses/LICENSE-2.0)
[![Maven Central](https://img.shields.io/maven-central/v/org.tkit.maven/tkit-mp-openapi-plugin?logo=java&style=for-the-badge)](https://maven-badges.herokuapp.com/maven-central/org.tkit.maven/tkit-mp-openapi-plugin)
[![GitHub Actions Status](<https://img.shields.io/github/workflow/status/1000kit/tkit-mp-openapi-plugin/build?logo=GitHub&style=for-the-badge>)](https://github.com/1000kit/tkit-mp-openapi-plugin/actions/workflows/build.yml)

## What it does

This plugin allows you to generate Openapi schemas from code (JaxRS/Quarkus). Quarkus does support Openapi schema generation out of the box, however with some limitations:  
* Schema generation happens at runtime 
* All API endpoints are merged into single schema 
* No possibility to exclude classes from the scan

This plugin fixes these issues by using the same underlying mechanism for schema generation(Smallrye Openapi) but allowing more granular control.

## Goal: generate

```xml
<plugin>
    <groupId>org.tkit.maven</groupId>
    <artifactId>tkit-mp-openapi-plugin</artifactId>
    <version>1.0.0</version>
    <executions>
        <execution>
            <id>generate</id>
            <goals>
                <goal>generate</goal>
            </goals>
            <configuration>
                <verbose>true</verbose>
                <classesDir>${project.build.outputDirectory}</classesDir>
                <!-- Extra MP config file that should be used as config source during the generator run -->
                <configFile>src/main/my.properties</configFile>
                <configFileOrdinal>200</configFileOrdinal>
                <properties>
                    <!-- Use this to control which packages should be included in scan -->
                    <mp.openapi.scan.exclude.packages>org.tkit.parameters.rs.external.v2</mp.openapi.scan.exclude.packages>
                </properties>
                <propertiesOrdinal>201</propertiesOrdinal>
                <format>YAML</format>
                <rootPath>root-rs</rootPath>
                <!-- where do you want to store the output -->
                <outputFile>${project.build.directory}/openapi.yaml</outputFile>
            </configuration>
        </execution>
    </executions>
</plugin>
```

If you want to generate multuiple schema files (for example internal.yaml and public-v1.yaml) then simply add 2 (or multiple) plugin executions with appropriate config.

### Parameters

|  Name | Default  | Values | Description  |
|---|---|---|---|
| verbose  | false | | The verbose flag  |
| classDir | ${project.build.outputDirectory} | | Directory of the classes |
| configFile | | | The micro-profile configuration property file |
| configFileOrdinal | 200 | | The micro-profile configuration property file ordinal number |
| properties | | | The micro-profile configuration properties in the plugin configuration |
| propertiesOrdinal | 201 | | The micro-profile configuration properties ordinal number |
| format | YAML | YAML, JSON | The output format |
| outputFile | ${project.build.directory}/openapi.yaml | | The output openAPI file |
| rootPath | | | When quarkus.http.root-path is set, it is useful to have rootPath also in output file  |

## Release

### Create a release

```bash
mvn semver-release:release-create
```

### Create a patch branch
```bash
mvn semver-release:patch-create -DpatchVersion=x.x.0
```


CHANGELOG.md
# Changelog

# 1.0.0

* Update all deps
* Sync with Quarkus 2.x

src/main/java/org/jboss/jandex/MergeIndexer.java
package org.jboss.jandex;

import java.io.InputStream;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MergeIndexer {

    public static String INDEX = "META-INF/jandex.idx";

    Map<DotName, List<AnnotationInstance>> annotations = new HashMap<>();
    Map<DotName, List<ClassInfo>> subclasses = new HashMap<>();
    Map<DotName, List<ClassInfo>> implementors = new HashMap<>();
    Map<DotName, ClassInfo> classes = new HashMap<>();

    public void addIndex(Index index) {
        annotations.putAll(index.annotations);
        subclasses.putAll(index.subclasses);
        implementors.putAll(index.implementors);
        classes.putAll(index.classes);
    }
    public Index complete() {
        return Index.create(annotations, subclasses, implementors, classes);
    }

    public void loadFromUrl(URL url) throws Exception {
        try (InputStream input = url.openStream()) {
            IndexReader reader = new IndexReader(input);
            Index index = reader.read();
            addIndex(index);
        }
    }
}


src/main/java/org/tkit/maven/mp/openapi/GenerateOpenApiMojo.java
/*
 * Copyright 2019 1000kit.org.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tkit.maven.mp.openapi;

import io.smallrye.config.PropertiesConfigSource;
import io.smallrye.config.SmallRyeConfig;
import io.smallrye.config.SmallRyeConfigBuilder;
import io.smallrye.openapi.api.OpenApiConfig;
import io.smallrye.openapi.api.OpenApiConfigImpl;
import io.smallrye.openapi.api.OpenApiDocument;
import io.smallrye.openapi.api.models.PathsImpl;
import io.smallrye.openapi.api.util.ClassLoaderUtil;
import io.smallrye.openapi.runtime.OpenApiProcessor;
import io.smallrye.openapi.runtime.io.Format;
import io.smallrye.openapi.runtime.io.OpenApiSerializer;
import org.apache.commons.lang3.StringUtils;
import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecution;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.plugins.annotations.ResolutionScope;
import org.apache.maven.project.MavenProject;
import org.apache.maven.shared.utils.io.DirectoryScanner;
import org.eclipse.microprofile.config.Config;
import org.eclipse.microprofile.config.ConfigProvider;
import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
import org.eclipse.microprofile.config.spi.ConfigSource;
import org.eclipse.microprofile.openapi.models.OpenAPI;
import org.eclipse.microprofile.openapi.models.PathItem;
import org.eclipse.microprofile.openapi.models.Paths;
import org.jboss.jandex.ClassInfo;
import org.jboss.jandex.Index;
import org.jboss.jandex.Indexer;
import org.jboss.jandex.MergeIndexer;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.*;

/**
 * Generate the openAPI file.
 */
@Mojo(name = "generate",
        defaultPhase = LifecyclePhase.PROCESS_CLASSES,
        requiresProject = true,
        requiresDependencyResolution = ResolutionScope.RUNTIME,
        aggregator = false,
        threadSafe = true)
public class GenerateOpenApiMojo extends AbstractMojo {

    /**
     * The classes directory.
     */
    @Parameter(defaultValue = "${project.build.outputDirectory}")
    private File classesDir;

    /**
     * The configuration file.
     */
    @Parameter(name = "configFile")
    private File configFile;

    /**
     * The configuration file ordinal number.
     */
    @Parameter(name = "configFileOrdinal", defaultValue = "200", required = true)
    private String configFileOrdinal;

    /**
     * The output file.
     */
    @Parameter(name = "outputFile", defaultValue = "${project.build.directory}/openapi.yaml", required = true)
    private File outputFile;

    /**
     * The format: YAML, JSON
     */
    @Parameter(name = "format", defaultValue = "YAML", required = true)
    private String format;

    /**
     * The verbose flag.
     */
    @Parameter(defaultValue = "false")
    private boolean verbose;

    /**
     * The micro-profile properties defined in the plugin configuration.
     */
    @Parameter(name = "properties")
    private Properties properties;

    /**
     * The micro-profile properties defined in the plugin configuration ordinal number.
     */
    @Parameter(name = "propertiesOrdinal", defaultValue = "201", required = true)
    private String propertiesOrdinal;

    /**
     * The mojo execution
     */
    @Parameter(defaultValue = "${mojoExecution}", readonly = true)
    private MojoExecution mojoExecution;

    @Parameter(defaultValue = "${project}", readonly = true, required = true)
    protected MavenProject project;

    @Parameter(name = "rootPath")
    private String rootPath;

    /**
     * {@inheritDoc }
     */
    @Override
    public void execute() throws MojoExecutionException, MojoFailureException {
        try {
            Index indexer = createIndexer();
            if (indexer == null) {
                return;
            }

            OpenApiConfig openApiConfig = createConfig();
            OpenAPI annotationModel = OpenApiProcessor.modelFromAnnotations(openApiConfig, indexer);

            OpenApiDocument document = OpenApiDocument.INSTANCE;
            document.reset();
            document.config(openApiConfig);
            document.modelFromAnnotations(annotationModel);
            document.filter(OpenApiProcessor.getFilter(openApiConfig, ClassLoaderUtil.getDefaultClassLoader()));
            document.initialize();

            releaseConfig();

            Format f = Format.YAML;
            if (Format.JSON.name().equals(format)) {
                f = Format.JSON;
            }
            appendRootPathIfExists(document);

            String output = OpenApiSerializer.serialize(document.get(), f);
            if (verbose) {
                getLog().info("OpenApi:\n" + output);
            }
            getLog().info("OpenAPI file in format " + format + " created: " + outputFile);

            // create the target directory
            Files.createDirectories(outputFile.toPath().getParent());

            // write the openapi to the output file
            Files.write(outputFile.toPath(), output.getBytes(StandardCharsets.UTF_8));

        } catch (Exception ex) {
            throw new MojoExecutionException("Error execute the plugin ", ex);
        }
    }

    private void appendRootPathIfExists(OpenApiDocument document) {
        if (StringUtils.isNotBlank(rootPath)) {
            Paths paths = document.get().getPaths();
            Iterator<String> iterator = paths.getPathItems().keySet().iterator();
            Paths updatedPaths = new PathsImpl();
            while (iterator.hasNext()) {
                String key = iterator.next();
                updatedPaths.addPathItem("/" + rootPath + key, paths.getPathItem(key));
            }
            document.get().setPaths(updatedPaths);
        }
    }

    /**
     * Creates the micro-profile configuration for the openAPI
     *
     * @return the openAPI configuration.
     * @throws MojoExecutionException if the method fails.
     */
    private OpenApiConfig createConfig() throws MojoExecutionException {
        Config config = ConfigProvider.getConfig();
        SmallRyeConfigBuilder cb = new SmallRyeConfigBuilder();

        config.getConfigSources().iterator().forEachRemaining(cb::withSources);

        // config file
        if (configFile != null && configFile.exists() && configFile.isFile()) {
            Properties prop = new Properties();
            try (InputStream input = new FileInputStream(configFile)) {
                prop.load(input);
            } catch (Exception e) {
                throw new MojoExecutionException("Error loading the properties " + configFile, e);
            }
            prop.setProperty(ConfigSource.CONFIG_ORDINAL, configFileOrdinal);
            cb.withSources(new PropertiesConfigSource(prop, configFile.getName()));
            getLog().info("Add the configuration source " + configFile);
        }

        // properties in the plugin configuration
        if (properties != null && !properties.isEmpty()) {
            properties.setProperty(ConfigSource.CONFIG_ORDINAL, propertiesOrdinal);
            String id = mojoExecution.getMojoDescriptor().getId() + " (" + mojoExecution.getExecutionId() + ")";
            cb.withSources(new PropertiesConfigSource(properties, id));
            getLog().info("Add the configuration source from maven plugin: " + id);
        }

        return new OpenApiConfigImpl(cb.build());
    }

    /**
     * Releases the micro-profile configuration for the openAPI
     * This is needed in order to have separate configuration properties for each plugin execution
     */
    private void releaseConfig() {
        Config config = ConfigProvider.getConfig();
        ConfigProviderResolver.instance().releaseConfig(config);
    }

    /**
     * Creates the indexer
     *
     * @return the indexer
     * @throws MojoExecutionException if the method fails.
     */
    private Index createIndexer() throws MojoExecutionException {

        if (!classesDir.exists()) {
            getLog().info("Directory does not exist! Directory: " + classesDir);
            return null;
        }

        final MergeIndexer indexer = new MergeIndexer();
        // load from dependencies
        loadIndexFromDependencies(indexer);
        // load from project
        loadIndexFromProject(indexer);
        // create index
        return indexer.complete();
    }

    private void loadIndexFromProject(final MergeIndexer indexer) throws MojoExecutionException {
        final DirectoryScanner scanner = new DirectoryScanner();
        scanner.setBasedir(classesDir);
        scanner.setIncludes("**/*.class");
        scanner.scan();
        final String[] files = scanner.getIncludedFiles();

        final Indexer tmpIndex = new Indexer();
        for (final String file : files) {
            if (file.endsWith(".class")) {
                File tmp = new File(classesDir, file);
                if (verbose) {
                    getLog().info("File: " + tmp);
                }
                try (FileInputStream fis = new FileInputStream(tmp)) {
                    final ClassInfo info = tmpIndex.index(fis);
                    if (verbose && info != null) {
                        getLog().info("Indexed " + info.name() + " (" + info.annotations().size() + " annotations)");
                    }
                } catch (final Exception e) {
                    throw new MojoExecutionException(e.getMessage(), e);
                }
            }
        }
        indexer.addIndex(tmpIndex.complete());
    }

    private void loadIndexFromDependencies(final MergeIndexer indexer) throws MojoExecutionException {
        try {
            List<String> elements = project.getRuntimeClasspathElements();
            if (elements != null && !elements.isEmpty()) {
                List<URL> tmp = new ArrayList<>();
                for (String element : elements) {
                    tmp.add(new File(element).toURI().toURL());
                }
                URL[] runtimeUrls = tmp.toArray(new URL[0]);
                URLClassLoader newLoader = new URLClassLoader(runtimeUrls, Thread.currentThread().getContextClassLoader());

                Enumeration<URL> items = newLoader.getResources(MergeIndexer.INDEX);
                while (items.hasMoreElements()) {
                    URL url = items.nextElement();
                    if (verbose) {
                        getLog().info("Dependency: " + url);
                    }
                    indexer.loadFromUrl(url);
                }
            }
        } catch (Exception e) {
            throw new MojoExecutionException("Error loading index from libraries.", e);
        }
    }
}
